File: tests/test_class_sh_void_ptr_capsule.cpp
#include <pybind11/smart_holder.h>

#include "pybind11_tests.h"

#include <memory>

namespace pybind11_tests {
namespace class_sh_void_ptr_capsule {

struct Valid {};

struct NoConversion {};

struct NoCapsuleReturned {};

struct AsAnotherObject {};

// Create a void pointer capsule for test. The encapsulated object does not
// matter for this test case.
PyObject *create_test_void_ptr_capsule() {
    static int just_to_have_something_to_point_to = 0;
    return PyCapsule_New(static_cast<void *>(&just_to_have_something_to_point_to), "int", nullptr);
}

int get_from_valid_capsule(const Valid *) { return 1; }

int get_from_shared_ptr_valid_capsule(const std::shared_ptr<Valid> &) { return 2; }

int get_from_unique_ptr_valid_capsule(std::unique_ptr<Valid>) { return 3; }

int get_from_no_conversion_capsule(const NoConversion *) { return 4; }

int get_from_no_capsule_returned(const NoCapsuleReturned *) { return 5; }

// https://github.com/pybind/pybind11/issues/3788
struct TypeWithGetattr {
    TypeWithGetattr() = default;
    int get_42() const { return 42; }
};

// https://github.com/pybind/pybind11/issues/3804
struct Base1 {
    int a1{};
};
struct Base2 {
    int a2{};
};

struct Base12 : Base1, Base2 {
    int foo() const { return 0; }
};

struct Derived1 : Base12 {
    int bar() const { return 1; }
};

struct Derived2 : Base12 {
    int bar() const { return 2; }
};

struct UnspecBase {
    virtual ~UnspecBase() = default;
    virtual int Get() const { return 100; }
};

int PassUnspecBase(const UnspecBase &sb) { return sb.Get() + 30; }

struct UnspecDerived : UnspecBase {
    int Get() const override { return 200; }
};

} // namespace class_sh_void_ptr_capsule
} // namespace pybind11_tests

using namespace pybind11_tests::class_sh_void_ptr_capsule;

PYBIND11_SMART_HOLDER_TYPE_CASTERS(Valid)
PYBIND11_SMART_HOLDER_TYPE_CASTERS(TypeWithGetattr)
PYBIND11_SMART_HOLDER_TYPE_CASTERS(Base1)
PYBIND11_SMART_HOLDER_TYPE_CASTERS(Base2)
PYBIND11_SMART_HOLDER_TYPE_CASTERS(Base12)
PYBIND11_SMART_HOLDER_TYPE_CASTERS(Derived1)
PYBIND11_SMART_HOLDER_TYPE_CASTERS(Derived2)
PYBIND11_SMART_HOLDER_TYPE_CASTERS(UnspecBase)
PYBIND11_SMART_HOLDER_TYPE_CASTERS(UnspecDerived)

TEST_SUBMODULE(class_sh_void_ptr_capsule, m) {
    py::classh<Valid>(m, "Valid");

    m.def("get_from_valid_capsule", &get_from_valid_capsule);
    m.def("get_from_shared_ptr_valid_capsule", &get_from_shared_ptr_valid_capsule);
    m.def("get_from_unique_ptr_valid_capsule", &get_from_unique_ptr_valid_capsule);
    m.def("get_from_no_conversion_capsule", &get_from_no_conversion_capsule);
    m.def("get_from_no_capsule_returned", &get_from_no_capsule_returned);
    m.def("create_test_void_ptr_capsule", []() {
        PyObject *capsule = create_test_void_ptr_capsule();
        return pybind11::reinterpret_steal<py::capsule>(capsule);
    });

    py::classh<TypeWithGetattr>(m, "TypeWithGetattr")
        .def(py::init<>())
        .def("get_42", &TypeWithGetattr::get_42)
        .def("__getattr__",
             [](TypeWithGetattr &, const std::string &key) { return "GetAttr: " + key; });

    py::classh<Base1>(m, "Base1");
    py::classh<Base2>(m, "Base2");

    py::classh<Base12, Base1, Base2>(m, "Base12")
        .def(py::init<>())
        .def("foo", &Base12::foo)
        .def("__getattr__",
             [](Base12 &, const std::string &key) { return "Base GetAttr: " + key; });

    py::classh<Derived1, Base12>(m, "Derived1").def(py::init<>()).def("bar", &Derived1::bar);

    py::classh<Derived2, Base12>(m, "Derived2").def(py::init<>()).def("bar", &Derived2::bar);

    py::classh<UnspecBase>(m, "UnspecBase");
    m.def("PassUnspecBase", PassUnspecBase);
    py::classh<UnspecDerived>(m, "UnspecDerived") // UnspecBase NOT specified as base here.
        .def(py::init<>())
        .def("as_pybind11_tests_class_sh_void_ptr_capsule_UnspecBase", [](UnspecDerived *self) {
            return py::reinterpret_steal<py::object>(
                PyCapsule_New(static_cast<void *>(self), nullptr, nullptr));
        });
}

File: tests/test_class_sh_void_ptr_capsule.py
import pytest

from pybind11_tests import class_sh_void_ptr_capsule as m


class Valid:
    def __init__(self):
        self.capsule_generated = False

    def as_pybind11_tests_class_sh_void_ptr_capsule_Valid(self):
        self.capsule_generated = True
        return m.create_test_void_ptr_capsule()


class NoConversion:
    def __init__(self):
        self.capsule_generated = False


class NoCapsuleReturned:
    def __init__(self):
        self.capsule_generated = False

    def as_pybind11_tests_class_sh_void_ptr_capsule_NoCapsuleReturned(
        self,
    ):
        pass


class AsAnotherObject:
    def __init__(self):
        self.capsule_generated = False

    def as_pybind11_tests_class_sh_void_ptr_capsule_Valid(self):
        self.capsule_generated = True
        return m.create_test_void_ptr_capsule()


@pytest.mark.parametrize(
    ("ctor", "caller", "expected"),
    [
        (Valid, m.get_from_valid_capsule, 1),
        (AsAnotherObject, m.get_from_valid_capsule, 1),
    ],
)
def test_valid_as_void_ptr_capsule_function(ctor, caller, expected):
    obj = ctor()
    assert caller(obj) == expected
    assert obj.capsule_generated


@pytest.mark.parametrize(
    ("ctor", "caller"),
    [
        (NoConversion, m.get_from_no_conversion_capsule),
        (NoCapsuleReturned, m.get_from_no_capsule_returned),
    ],
)
def test_invalid_as_void_ptr_capsule_function(ctor, caller):
    obj = ctor()
    with pytest.raises(TypeError):
        caller(obj)
    assert not obj.capsule_generated


@pytest.mark.parametrize(
    ("ctor", "caller", "pointer_type", "capsule_generated"),
    [
        (AsAnotherObject, m.get_from_shared_ptr_valid_capsule, "shared_ptr", True),
        (AsAnotherObject, m.get_from_unique_ptr_valid_capsule, "unique_ptr", True),
    ],
)
def test_as_void_ptr_capsule_unsupported(ctor, caller, pointer_type, capsule_generated):
    obj = ctor()
    with pytest.raises(RuntimeError) as excinfo:
        caller(obj)
    assert pointer_type in str(excinfo.value)
    assert obj.capsule_generated == capsule_generated


def test_type_with_getattr():
    obj = m.TypeWithGetattr()
    assert obj.get_42() == 42
    assert obj.something == "GetAttr: something"


def test_multiple_inheritance_getattr():
    d1 = m.Derived1()
    assert d1.foo() == 0
    assert d1.bar() == 1
    assert d1.prop1 == "Base GetAttr: prop1"

    d2 = m.Derived2()
    assert d2.foo() == 0
    assert d2.bar() == 2
    assert d2.prop2 == "Base GetAttr: prop2"


def test_pass_unspecified_base():
    assert m.PassUnspecBase(m.UnspecDerived()) == 230
